Dart/src/com/jetbrains/lang/dart/ide/runner/DartExecutionHelper.java

```diff
@@ -0,0 +1,115 @@
+import com.intellij.openapi.fileEditor.OpenFileDescriptor;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.search.GlobalSearchScope;
+import com.jetbrains.lang.dart.analyzer.DartAnalysisServerService;
+import com.jetbrains.lang.dart.analyzer.DartServerData;
+import com.jetbrains.lang.dart.ide.errorTreeView.DartProblemsView;
+import com.jetbrains.lang.dart.resolve.DartResolveScopeProvider;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DartExecutionHelper {
+
+  private DartExecutionHelper() {
+  }
+
+  @SuppressWarnings("unused")
+  public static boolean hasIssues(@NotNull final Project project, @NotNull VirtualFile launchFile) {
+    return !getIssues(project, launchFile).isEmpty();
+  }
+
+  public static List<DartServerData.DartError> getIssues(@NotNull final Project project, @NotNull VirtualFile launchFile) {
+    return getIssues(project, launchFile, true);
+  }
+
+  public static List<DartServerData.DartError> getIssues(@NotNull final Project project,
+                                                         @NotNull VirtualFile launchFile,
+                                                         boolean onlyErrors) {
+    GlobalSearchScope scope = DartResolveScopeProvider.getDartScope(project, launchFile, true);
+    if (scope == null) {
+      return Collections.emptyList();
+    }
+
+    // Collect errors.
+    final DartAnalysisServerService analysisServerService = DartAnalysisServerService.getInstance(project);
+    List<DartServerData.DartError> errors = analysisServerService.getErrors(scope);
+    if (onlyErrors) {
+      errors = errors.stream().filter(DartServerData.DartError::isError).collect(Collectors.toList());
+    }
+
+    return errors;
+  }
+
+  @SuppressWarnings("unused")
+  public static void displayIssues(@NotNull final Project project, @NotNull VirtualFile launchFile) {
+    displayIssues(project, launchFile, "Analysis issues with launch", null);
+  }
+
+  @SuppressWarnings("SameParameterValue")
+  public static void displayIssues(@NotNull final Project project,
+                                   @NotNull VirtualFile launchFile,
+                                   @NotNull String launchTitle,
+                                   @Nullable Icon icon) {
+    clearIssueNotifications(project);
+
+    List<DartServerData.DartError> errors = getIssues(project, launchFile);
+
+    if (errors.isEmpty()) {
+      return;
+    }
+
+    final DartProblemsView problemsView = DartProblemsView.getInstance(project);
+
+    final String content;
+    if (errors.size() == 1) {
+      content = errors.get(0).getMessage() + " (<a href=\"issues\">show</a>)";
```

Comment:
That makes sense re: putting the link creation and handling behavior in one place.

Re: the `show` link, having fewer things in that notifications is better. If you can determine when the tool window is showing, sgtm.